internal/service/costsheet.go

```diff
@@ -0,0 +1,428 @@
+	GetServiceParams() ServiceParams
+}
+
+// costSheetService implements the CostSheetService interface.
+// It coordinates between different components to calculate costs and margins.
+type costsheetService struct {
+	ServiceParams
+}
+
+// NewCostSheetService creates a new instance of the cost sheet service with the required dependencies.
+func NewCostSheetService(params ServiceParams) CostSheetService {
+	return &costsheetService{
+		ServiceParams: params,
+	}
+}
+
+// GetInputCostForMargin implements the core business logic for calculating input costs.
+// The process involves:
+// 1. Getting subscription details and time period
+// 2. Processing fixed costs from subscription line items
+// 3. Processing usage-based costs by gathering meter data
+// 4. Aggregating total cost and item-specific costs
+func (s *costsheetService) GetInputCostForMargin(ctx context.Context, req *dto.GetCostBreakdownRequest) (*dto.CostBreakdownResponse, error) {
+	// service instances
+	subscriptionService := NewSubscriptionService(s.ServiceParams)
+	priceService := NewPriceService(s.PriceRepo, s.MeterRepo, s.Logger)
+	eventService := NewEventService(s.EventRepo, s.MeterRepo, s.EventPublisher, s.Logger, s.Config)
+
+	// Get subscription details first
+	sub, err := subscriptionService.GetSubscription(ctx, req.SubscriptionID)
+	if err != nil {
+		return nil, ierr.WithError(err).
+			WithMessage("failed to get subscription").
+			WithHint("failed to get subscription details").
+			Mark(ierr.ErrInternal)
+	}
+
+	// Determine time period - use request time if provided, otherwise use subscription period
+	var startTime, endTime time.Time
+	if req.StartTime != nil && req.EndTime != nil {
+		startTime = *req.StartTime
+		endTime = *req.EndTime
+	} else {
+		startTime = sub.Subscription.CurrentPeriodStart
+		endTime = sub.Subscription.CurrentPeriodEnd
+	}
+
+	var totalCost decimal.Decimal
+	itemCosts := make([]dto.CostBreakdownItem, 0)
+
+	// Create a map to track unique meters for usage-based pricing
+	uniqueMeters := make(map[string]struct{})
+
+	// First pass: Process fixed costs and collect meters for usage-based costs
+	for _, item := range sub.LineItems {
+		if item.Status != types.StatusPublished {
+			continue
+		}
+
+		// Get price details
+		priceResp, err := priceService.GetPrice(ctx, item.PriceID)
+		if err != nil {
+			return nil, ierr.WithError(err).
+				WithMessage("failed to get price").
+				WithHint("failed to get price").
+				Mark(ierr.ErrInternal)
+		}
+
+		if priceResp.Price.Type == types.PRICE_TYPE_FIXED {
+			// For fixed prices, use the price amount directly
+			cost := priceResp.Price.Amount.Mul(item.Quantity)
+			itemCosts = append(itemCosts, dto.CostBreakdownItem{
+				Cost: cost,
+			})
+			totalCost = totalCost.Add(cost)
+		} else if priceResp.Price.Type == types.PRICE_TYPE_USAGE && item.MeterID != "" {
+			// For usage-based prices, collect the meter ID for later processing
+			uniqueMeters[item.MeterID] = struct{}{}
```

Comment:
no need to dedupe here just in case there are more than 1 line item with same meter